ramips: add support to Strong 1200

vk496 opened the pull request:
Support for Strong 1200

[https://forum.openwrt.org/t/support-for-strong-1200/22768](https://forum.openwrt.org/t/support-for-strong-1200/22768)

Specification:
- SoC: MediaTek MT7621A (880 MHz)
- Flash: 16 MiB
- RAM: 128 MiB
- Wireless: 2.4Ghz(MT7602EN) and 5Ghz (MT7612EN)
- Ethernet speed: 10/100/1000
- Ethernet ports: 4+1
- 1x USB 3.0
- 1x microSD reader
- Serial baud rate of Bootloader and factory firmware: 57600

To flash via webinterface use first initramfs.bin and after (from the OpenWrt) the sysupgrade.bin

Some notes:
- some microSD not work (kernel bug?):

```
[  173.767665] mtk-sd 1e130000.sdhci: no support for card's volts
[  173.779318] mmc0: error -22 whilst initialising SDIO card
[  173.791941] mtk-sd 1e130000.sdhci: no support for card's volts
[  173.803598] mmc0: error -22 whilst initialising MMC card
[  173.877989] mtk-sd 1e130000.sdhci: no support for card's volts
[  173.889610] mmc0: error -22 whilst initialising SDIO card
[  173.902535] mtk-sd 1e130000.sdhci: card claims to support voltages below defined range
[  173.918337] mtk-sd 1e130000.sdhci: no support for card's volts
[  173.929975] mmc0: error -22 whilst initialising MMC card
[  173.998351] mtk-sd 1e130000.sdhci: no support for card's volts
[  174.009972] mmc0: error -22 whilst initialising SDIO card
[  174.023202] mtk-sd 1e130000.sdhci: no support for card's volts
[  174.034820] mmc0: error -22 whilst initialising MMC card
```

- Wifi LED (2.4 and 5) not available

blocktrron commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -359,6 +359,15 @@ define Device/sk-wb8
+	IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  DEVICE_TITLE := Strong Gigabit Dual Band 1200
+	UIMAGE_NAME:= WR1201_8_128
+  DEVICE_PACKAGES := kmod-sdhci-mt7620 kmod-mt76 kmod-usb3 kmod-usb-ledtrig-usbport wpad-mini
```
indentations are mixed here - should be two spaces only

vk496 replied:
> indentations are mixed here - should be two spaces only

Fixed!

blocktrron commented:
It would be great (but does not seem to be a requirement) if the commit message would include some information about the devices hardware (SoC / supported WiFI standards / RAM/ Flash) and a short description how to install OpenWRT on the device. 

Some examples:
https://github.com/openwrt/openwrt/commit/5543d63fc84e64be83afaa4241369a07797f041b
https://github.com/openwrt/openwrt/commit/adbd9d1f1f07f77b0f511027c1cfea87cefe47be
https://github.com/openwrt/openwrt/commit/3f8c5d547677e7f4e450e88379c03448b8f53b1d

vk496 commented:
> It would be great (but does not seem to be a requirement) if the commit message would include some information about the devices hardware (SoC / supported WiFI standards / RAM/ Flash) and a short description how to install OpenWRT on the device.
> 
> Some examples:
> [5543d63](https://github.com/openwrt/openwrt/commit/5543d63fc84e64be83afaa4241369a07797f041b)
> [adbd9d1](https://github.com/openwrt/openwrt/commit/adbd9d1f1f07f77b0f511027c1cfea87cefe47be)
> [3f8c5d5](https://github.com/openwrt/openwrt/commit/3f8c5d547677e7f4e450e88379c03448b8f53b1d)

Done!

vk496 commented:
Hello,

First, thank you for your time. I updated the branch with the latest commits from master and applied your modifications. I will post the results in the next message. Meanwhile, I would like to discuss some points:

### uImage header
It is possible to modify the uimage name ONLY for the **initramfs** image (without creating functions to manually edit the image and recalculate the CRC32)?

The problem is that for flashing the firmware from OEM, is mandatory to flash first the **initramfs** and after the **squashfs** (if you flash directly the **squashfs**, you [get bootloop](https://forum.openwrt.org/t/support-for-strong-1200/22768/19?u=vk496)), and the OEM webpage validate the image by uimage header name.

Using the `UIMAGE_NAME` do the trick, but it apply the name for both images.

### LEDs
Thanks for the tip with the leds. I will try you suggestion. I also created a [dedicated thread](https://forum.openwrt.org/t/why-some-mt7621-leds-are-not-working/23939) in the forum to understand better the issue and how to solve it. Do you have some extra references related to pinmux? It is not clear for me what they exactly do.

BR

mkresin commented:
> ### uImage header
>It is possible to modify the uimage name ONLY for the **initramfs** image (without creating functions to manually edit the image and recalculate the CRC32)?
>
>The problem is that for flashing the firmware from OEM, is mandatory to flash first the **initramfs** and after the **squashfs** (if you flash directly the **squashfs**, you [get bootloop](https://forum.openwrt.org/t/support-for-strong-1200/22768/19?u=vk496)), and the OEM webpage validate the image by uimage header name.
>
>Using the `UIMAGE_NAME` do the trick, but it apply the name for both images.

To my knowledge it is not possible to set different uImage names for initramfs and "normal" sysupgrade kernel.

What issues does it cause if the uimage name is "WR1201_8_128" for the kernel in the sysupgrade image? That fact that the uimage name is validated should be mentioned in the commit message.

vk496 commented:
> 
> To my knowledge it is not possible to set different uImage names for initramfs and "normal" sysupgrade kernel.
> 
> What issues does it cause if the uimage name is "WR1201_8_128" for the kernel in the sysupgrade image? That fact that the uimage name is validated should be mentioned in the commit message.

If the special header name is set only in the initramfs, it would be not possible to soft-brick the device during the OpenWrt flash from the OEM (the OEM webinterface would accept only the **initramfs**). I think is very useful for novel users.

Related, to the LEDS: It works!

Both WLAN2.4 and WLAN5 are connected to the wireless chips. Thanks for the tip! The rest of the OpenWrt main functions work properly (WLAN, LEDS, Switch, USB and microSD).

BR

mkresin commented:
> If the special header name is set only in the initramfs, it would be not possible to soft-brick the device during the OpenWrt flash from the OEM (the OEM webinterface would accept only the initramfs). I think is very useful for novel users.

Good point, but not possible yet.

Please drop the links to the forum from the commit message as I had done it with the commit in my staging tree. Please extend the commit message with the information that the OEM webinterface writes only as much bytes as listed in the uImage header field to the flash and that the OEM webinterface evaluates the name field of uImage header.

That should be what is left to merge the PR.

vk496 commented:
> Please drop the links to the forum from the commit message as I had done it with the commit in my staging tree. Please extend the commit message with the information that the OEM webinterface writes only as much bytes as listed in the uImage header field to the flash and that the OEM webinterface evaluates the name field of uImage header.

Done

> Good point, but not possible yet.

Already implemented it with some perl & Linux hacks :)

https://github.com/openwrt/openwrt/pull/1473/files#diff-078cf44df243ae35f6fa7395e9a95abaR105

Tried to flash from the OEM and works like a charm (the squashfs is refused and the initramfs is accepted)

mkresin commented:
> Already implemented it with some perl & Linux hacks :)
>
> https://github.com/openwrt/openwrt/pull/1473/files#diff-078cf44df243ae35f6fa7395e9a95abaR105

Please drop it, I'm not going to accept such hacks. It's utterly hard to read and understand. Both good indicators that it will be broken within a year.

Please try something similar to the following instead:

```
diff --git a/target/linux/ramips/image/mt7621.mk b/target/linux/ramips/image/mt7621.mk
index 4a5e99fb4c..04806db27f 100644
--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -71,6 +71,20 @@ define Build/ubnt-erx-factory-image
 	fi
 endef
 
+# The OEM webinterface expects an kernel with initramfs which has the uImage
+# header field ih_name set to "WR1201_8_128".
+# We don't wan't to set the header name field for the kernel include in the
+# sysupgrade image as well, as this image shouldn't be accepted by the OEM
+# webinterface. It will soft-brick the board.
+define Build/wr1201-factory-header
+	mkimage -A $(LINUX_KARCH) \
+		-O linux -T kernel \
+		-C lzma -a $(KERNEL_LOADADDR) -e $(if $(KERNEL_ENTRY),$(KERNEL_ENTRY),$(KERNEL_LOADADDR)) \
+		-n 'WR1201_8_128' -d $@ $@.new
+	mv $@.new $@
+endef
+
+
 define Device/11acnas
   DTS := 11ACNAS
   IMAGE_SIZE := $(ralink_default_fw_size_16M)
@@ -339,6 +353,7 @@ TARGET_DEVICES += mikrotik_rbm11g
 define Device/mtc_wr1201
   DTS := WR1201
   IMAGE_SIZE := 16000k
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | wr1201-factory-header
   DEVICE_TITLE := MTC Wireless Router WR1201
   DEVICE_PACKAGES := kmod-sdhci-mt7620 kmod-mt76x2 kmod-usb3 \
 	kmod-usb-ledtrig-usbport wpad-basic
```

mkresin commented:
> I would prefer to adapt the function and use it as generic, passing the name as parameter. Do you agree?

Please keep it in mt7621.mk. We don't need another function to pass the uimage name, as we have the variable UIMAGE_NAME for that purpose.

It is something only required for the wr1201 and can be as fixed as my example.

vk496 commented:
Okay, will use directly your patch

vk496 commented:
Ready!

mkresin commented:
> Edit: I remembered. There are more devices of the same company (N300 and N750). Probably they will need exactly the same fix with different name

We will take care of these as soon as support for these will be added.

> Ready!

No, the factory-header command now receives a parameter which isn't at the moment required.

You make it really hard to bring the PR to an end. I will fix it now on my own before pushing it. Enough time wasted.

vk496 committed:
ramips: add support for MTC Wireless Router WR1201

MTC Wireless Router WR1201 is the OEM name of the board. It is also sold
rebranded as STRONG Dual Band Gigabit Router 1200.

Specification:
- SoC: MediaTek MT7621A (880 MHz)
- Flash: 16 MiB
- RAM: 128 MiB
- Wireless: 2.4Ghz(MT7602EN) and 5Ghz (MT7612EN)
- Ethernet speed: 10/100/1000
- Ethernet ports: 4+1
- 1x USB 3.0
- 1x microSD reader
- Serial baud rate of Bootloader and factory firmware: 57600

The OEM webinterface writes only as much bytes as listed in the
uImage header field to the flash. Also, the OEM webinterface
evaluates the name field of uImage header before flashing (the
string "WR1201_8_128")

To flash via webinterface, is mandatory to use first initramfs.bin
and after (from the OpenWrt) the sysupgrade.bin

Some notes:
- Some microSD will not work:

  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising SDIO card
  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising MMC card
  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising SDIO card
  mtk-sd 1e130000.sdhci: card claims to support voltages below defined range
  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising MMC card
  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising SDIO card
  mtk-sd 1e130000.sdhci: no support for card's volts
  mmc0: error -22 whilst initialising MMC card

Signed-off-by: Valentín Kivachuk <vk18496@gmail.com>

vk496 commented:
I would prefer the more generic way, but ok. Fixed to yout original func.

Thnks for your time.

BR

mkresin commented:
Merged, thanks!
